src/vs-picgo/index.ts

```diff
@@ -0,0 +1,192 @@
+import * as fs from 'fs';
+import * as path from 'path';
+import * as os from 'os';
+import * as vscode from 'vscode';
+
+import PicGo from 'picgo/dist/core/PicGo';
+import { ImgInfo } from 'picgo/dist/utils/interfaces';
+
+import { promisify } from 'util';
+
+import { formatParam, formatString } from '../utils';
+
+const _ = require('lodash');
+const _db = require('lodash-id');
+_.mixin(_db);
+
+const writeFileP = promisify(fs.writeFile);
+const readFileP = promisify(fs.readFile);
+
+export interface INotice {
+  body: string;
+  text: string;
+  title: string;
+}
+
+export interface IUploadName {
+  date: string;
+  dateTime: string;
+  fileName: string;
+  extName: string;
+  mdFileName: string;
+  [key: string]: string;
+}
+
+export interface IOutputUrl {
+  uploadedName: string;
+  url: string;
+  [key: string]: string;
+}
+
+export default class VSPicgo {
+  private dataPath: string;
+  private editor: vscode.TextEditor;
+  private picgo: PicGo;
+
+  constructor(editor: vscode.TextEditor) {
+    const picgoConfigPath = vscode.workspace.getConfiguration('picgo').get<string>('configPath');
+    if (picgoConfigPath) {
+      this.picgo = new PicGo(picgoConfigPath);
+    } else {
+      this.picgo = new PicGo();
+      const picBed = vscode.workspace.getConfiguration('picgo.picBed');
+      this.picgo.setConfig({ picBed });
+    }
+    this.editor = editor;
+    this.dataPath = this.getDataPath();
+    // Before upload, we change names of the images.
+    this.addRenameListener();
+    // After upload, we use the custom output format.
+    this.addGenerateOutputListener();
+  }
+
+  addGenerateOutputListener() {
+    this.picgo.on('finished', async (ctx: PicGo) => {
+      let urlText = '';
+      const outputFormatTemplate =
+        vscode.workspace.getConfiguration('picgo').get<string>('customOutputFormat') || '![${uploadedName}](${url})';
+      try {
+        urlText = ctx.output.reduce((acc: string, cur: ImgInfo): string => {
+          return `${acc}${formatString(outputFormatTemplate, {
+            uploadedName: cur.fileName || '',
+            url: cur.imgUrl,
+          })}\n`;
+        }, '');
+        urlText = urlText.trim();
+        await this.updateLog(ctx.output);
+      } catch (err) {
+        if (err instanceof SyntaxError) {
+          vscode.window.showErrorMessage(
+            `The data file ${this.dataPath} has syntax error, ` +
+              `please fix the error by yourself or delete the data file and vs-picgo will recreate for you.`,
+          );
+        } else {
+          vscode.window.showErrorMessage(`Failed to read from data file ${this.dataPath}: ${err || ''}`);
+        }
+        return;
+      }
+      this.editor.edit(textEditor => {
+        textEditor.replace(this.editor.selection, urlText);
+        vscode.window.showInformationMessage('Upload successfully');
+      });
+    });
+  }
+
+  addRenameListener() {
+    this.picgo.on('beforeUpload', (ctx: PicGo) => {
```

Comment:
here, register a `beforeUploadPlugins` would be safer.
For example picgo's [uploader](https://github.com/Molunerfinn/Picgo/blob/dev/src/main/utils/uploader.js#L92-L113).
Because listening to `beforeUpload` event, there is no guarantee that the operation will be completed before uploading. While `beforeUploadPlugins` will be called before uploading.

Reply:
After some test, I found #37 is related to this, for example, if I simply change the fileName to `lalala`:
```js
this.picgo.on('beforeUpload', (ctx: PicGo) => {
     
      // ...
        ctx.output[0].fileName = 'lalala';
        console.log(`ctx.output[0].fileName = ${ctx.output[0].fileName}`)
     // ...
    });
```

The result we get looks like this:

![image](https://user-images.githubusercontent.com/24741764/56664986-c45fe280-66db-11e9-878f-43bf8e8116bf.png)

This is really strange, I will try to solve it by registering a `beforeUploadPlugins`.

Reply:
Oh, it's my fault, I forgot the `path`:

https://github.com/PicGo/PicGo-Core/blob/904804d3e0c0fde4900d34c9b74fdc955735b705/src/plugins/uploader/tcyun.ts#L50

Reply:
I have fixed #37 in dev, after we merge this PR, we should update the description of each pic bed in settings according to https://picgo.github.io/PicGo-Core-Doc/zh/guide/config.html, too. @Spades-S